### PR TITLE
Add new usrdat modes for CLM/MOSART and MOSART

### PR DESCRIPTION
**Pull request by apcraig**
- Adds new usrdat resolutions for ELM/MOSART and MOSART.
- Removes RMOSARTNLDAS compset because it's not working
- Adds four new tests

[BFB]

### CODE REVIEW

**Comment by bishtgautam** [comment]
From @apcraig

> 
> Add new usrdat modes for CLM/MOSART and MOSART.
> 
> - Remove compset RMOSARTNLDAS because it's not working
> - Add compsets RMOSQIAN, RMOSGPCC.  Tested with r05_r05.  Verified the following configurations,
>   - --compset INLDASCNPECACNTBC --res NLDAS_NLDAS
>   - --compset ICLM45 --res r05_r05
>   -  --compset RMOSGPCC --res r05_r05 
>   -  --compset RMOSQIAN --res r05_r05
> - Add support for new USRDAT modes.  The following were tested, CLM_USRDAT, CLMMOS_USRDAT, MOS_USRDAT.  There are sample test scripts on compy,
>    - /qfs/people/crai556/CLM_USRDAT.01
>    - /qfs/people/crai556/CLMMOS_USRDAT.01
>    - /qfs/people/crai556/MOS_USRDAT.01
> - Add QIAN and GPCC dlnd data modes.
> - Tested e3sm_land_developer suite versus 650241f0db5, all tests pass except the following (which is probably just machine performance noise).
> 
> >     FAIL ERS.f09_g16.I1850CLM45CN.compy_intel TPUTCOMP Error: Computation time increase > 5 pct from baseline
> >     FAIL ERS.f19_g16.I1850CLM45.compy_intel.clm-betr TPUTCOMP Error: Computation time increase > 5 pct from baseline
> >     FAIL ERS.ne11_oQU240.I20TRCLM45.compy_intel TPUTCOMP Error: Computation time increase > 5 pct from baseline
> >     FAIL SMS_Ly2_P1x1.1x1_smallvilleIA.ICLM45CNCROP.compy_intel.clm-lulcc_sville TPUTCOMP Error: Computation time increase > 5 pct from baseline
> > 
> 
> - No new tests were added, but I recommend adding some new tests for the RM r05_r05 configurations.
> - In the future, there are several things that should be done to improve the implementation
>   - CLM_USRDAT should not be the component gridname, it should be USRDAT
>   - There should be just one USRDAT mode that sets the component grid to USRDAT for ALL components.  Then you just need a subset of components that support USRDAT mode and a set of compsets that work with USRDAT mode.  There is NO REASON to implement a new *_USRDAT grid for every combination of components you want to run in USRDAT mode.  In fact, if you look at the implementation, CLMMOS_USRDAT and MOS_USRDAT do exactly the same thing and are interchangeable with the compset they work with.  Again, this is silly, but to fix this would mean cleaning up a few of other things.
>   - If this were done more generally, you would not have to support "null" grid mode in any component
> 

@apcraig, In E3SM, the first comment of each PR is copied/pasted when the PR is merged to next and master. So, the first comment is a succinct description of the PR. Thus, I have modified your first comment.

File: cime/src/components/data_comps_mct/dlnd/cime_config/namelist_definition_dlnd.xml [hunk 1]
```diff
@@ -138,6 +154,8 @@
     <values>
       <value stream="sno.cplhist">$DLND_CPLHIST_CASE.cpl.hl2x1yr_glc.%y-01-01.nc</value>
       <value stream="clm.nldas">mosart_nldas_hist_1975.clm2.h1.1975-01-01-00000.nc</value>
+      <value stream="lnd.qian">QIAN.daily.1948-2004.nc</value>
+      <value stream="lnd.gpcc">GPCC.daily.nc</value>
```
Comment (hydrotian) [inline cime/src/components/data_comps_mct/dlnd/cime_config/namelist_definition_dlnd.xml:158]: Do we need to upload these two nc files to the server or they are already there?

Comment (bishtgautam) [reply cime/src/components/data_comps_mct/dlnd/cime_config/namelist_definition_dlnd.xml:158]: @hydrotian, @apcraig, I missed this comment and have merged the PR to next but I'm going to revert it. Where are these files on Compy and do we need these new files?

Comment (apcraig) [reply cime/src/components/data_comps_mct/dlnd/cime_config/namelist_definition_dlnd.xml:158]: Both data files are on compy's inputdata area,

/compyfs/inputdata/lnd/dlnd7/hcru_hcru/

I do not know whether these files have been added to the inputdata server or whether they should be.

Comment (hydrotian) [reply cime/src/components/data_comps_mct/dlnd/cime_config/namelist_definition_dlnd.xml:158]: @bishtgautam I guess you may not need them to pass the tests, but these are default runoff forcing data for the new usrdat functions.

**Comment by rljacob** [comment]
The tests are complaining they can't find input data. 
`Model dlnd missing file domainfile = '/share/domains/domain.lnd.r05_oEC60to30v3.190418.nc'`

The leading "/" should be removed from their entry in the namelists.

**Comment by hydrotian** [comment]
@bishtgautam I can see this file is in compy:/compyfs/inputdata/share/domains. Could you please copy it to the machine that is complaining?

**Comment by rljacob** [comment]
The problem is this line in the shell_commands in the mos-usrdat testmod:
`./xmlchange LND_DOMAIN_PATH=$DIN_LOC_ROOT/share/domains`

$DIN_LOC_ROOT isn't defined so the path winds up starting with /share which doesn't exist.

**Comment by rljacob** [comment]
@bishtgautam to fix you need to add ./xmlquery to shell_commands to get the value of DIN_LOC_ROOT.

For the gpcc_1972 testmod, this file is missing from the server:  inputdata/lnd/dlnd7/hcru_hcru/GPCC.daily.nc

For the test:  ERS.CLMMOS_USRDAT.ICLM45.sandiatoss3_intel.clm-clmmos_usrdat, the testmod must have a typo because it can not be found.  It resolves to: clm/cime_config/testdefs/testmods_dirs/clm/clmmos_usrdat

For the test: ERS.r05_r05.RMOSQIAN.sandiatoss3_intel.mosart-qian_1972  this file is missing from the server:  inputdata/lnd/dlnd7/hcru_hcru/QIAN.daily.1948-2004.nc

**Comment by rljacob** [comment]
@hydrotian its not necessary to copy input files to each machine.  They are placed on our data server and then CIME downloads them to each machine as needed.  Provided the path is correct.

**Comment by hydrotian** [comment]
@rljacob thanks for explaining.

**Comment by bishtgautam** [comment]
@robjacob, For fixing the DIN_LOC_ROOT, `./xmlchange LND_DOMAIN_PATH="\$DIN_LOC_ROOT/share/domains"` should work

**Comment by rljacob** [comment]
Be sure to test it yourself with create_test before pushing a fix.

**Comment by rljacob** [comment]
@apcraig Hi Tony.  Just now looking at these USRDAT modes.  What was the reason for creating them?  Does using USRDAT as a grid type mean that one MUST then define the domains/grids in a usermod ?

**Comment by bishtgautam** [comment]
>Does using USRDAT as a grid type mean that one MUST then define the domains/grids in a usermod ?

Yes, it is for a user-defined grid. The test mods need to define a domain/grid. We are using the same thing for ELM tests with USRDAT in which the [testmods](https://github.com/E3SM-Project/E3SM/tree/master/components/elm/cime_config/testdefs/testmods_dirs/elm/usrdat) define the `<LND/ATM>_DOMAIN_<FILE/PATH>` and ELM `fsurdat`.

**Pull request merged**